### PR TITLE
Goldilocks implementation in Rust

### PR DESCRIPTION
**Pull request by 3lkn**
This is the Rust implementation of Goldilocks; for analogous changes in C++, see [@jbruestle's working branch](https://github.com/risc0/risc0/blob/jeremy-goldilocks/risc0/zkp/core/fp.h). 

**Special review request: Comments overlap heavily with baby_bear.rs, looking for comment review from a documentation perspective so we don't have competing explanation variants.**

Notes:
* Removed options to create field elements from u32 values; may restore this for test purposes somewhere down the (generic) line.

Testing notes:
* Added granular tests for arithmetic operations. 
* In the tests, to generate expected values for field operations, most operations are performed in 128-bit space and then reduced mod P. 
* Because we don't have the 'baby bear' promise not to overflow addition, subtraction is a little case-dependent, and I'd welcome a simplification from any low-level math wizards.

### CODE REVIEW

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 1]
```diff
@@ -0,0 +1,745 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// ! Goldilocks field.
+/// ! Support for the base finite field modulo 2^64 - 2^32 + 1
+use crate::field::{self, Elem as FieldElem};
+
+use core::ops;
+
+use bytemuck::{Pod, Zeroable};
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:21]: The sorting of groups should go:

* std/core/alloc (stuff from rust)
* third party (i.e. bytemuck)
* crate::*

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 2]
```diff
@@ -0,0 +1,745 @@
...
+}
+
+macro_rules! rou_array {
+    [$($x:literal),* $(,)?] => {
+        [$(Elem::new($x)),* ]
+    }
+}
+
+/// Maximum power of two for which we have a root of unity using Goldilocks
+/// field
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:90]: This comment applies specifically to `MAX_ROU_PO2`. IIUC, `ROU_FWD` and `ROU_REV` are powers of our canonical 32nd RoU.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 3]
```diff
@@ -0,0 +1,745 @@
...
+        }
+        Elem::from(val)
+    }
+}
+
+macro_rules! rou_array {
+    [$($x:literal),* $(,)?] => {
+        [$(Elem::new($x)),* ]
+    }
+}
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:87]: What's the purpose of this? Why can't we just create these directly?

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+    // Get two high words
+    let med: u32 = (prod >> 64) as u32;
+    let high: u32 = (prod >> 96) as u32;
+    // Subtract out high bits, add in P if underflow
+    if ret >= (high as u64) {
+        ret = ret.wrapping_sub(high as u64);
+    } else {
+        ret = ret.wrapping_sub(high as u64);
+        ret = ret.wrapping_add(P);
+    }
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:278]:
```suggestion
    let ret = 
    if ret >= (high as u64) {
        ret.wrapping_sub(high as u64)
    } else {
        ret.wrapping_sub(high as u64).wrapping_add(P)
    }
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+    return if x > lhs { x.wrapping_add(P) } else { x };
+}
+
+/// Wrapping multiplication of [Elem] using Goldilocks prime
+fn mul(lhs: u64, rhs: u64) -> u64 {
+    // To prevent u64 overflow, we first perform with u128
+    let prod: u128 = (lhs as u128).wrapping_mul(rhs as u128);
+
+    // Initialize result to low 64 bits (by converting)
+    let mut ret: u64 = prod as u64;
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:268]: In rust, you can 'shadow' variables without ambiguity. So instead of using a `mut`, you could just create a new lvalue for each.

For example, you could say:
```
let mut x = 1;
x = x + 1;
```

Or:
```
let x = 1;
let x = x + 1;
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+        ret = ret.wrapping_add(P);
+    }
+    // Compute shifted effect of medium
+    let med_shift = ((med as u64) << 32).wrapping_sub(med as u64);
+    // Add in, if overflow, subtract a P
+    ret = ret.wrapping_add(med_shift);
+    if ret < med_shift || ret >= P {
+        ret = ret.wrapping_sub(P)
+    }
+    ret as u64
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:286]: It's not clear why we need to cast at the end. `ret` should be a `u64` at this point.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+const EXT_SIZE: usize = 2;
+
+/// Instances of `ExtElem` are elements of a finite field `F_p^2`. They are
+/// represented as elements of `F_p[X] / (X^2 - 11)`. This large
+/// finite field (about `2^128` elements) is used when the security of
+/// operations depends on the size of the field. The field extension `ExtElem`
+/// has `Elem` as a subfield, so operations on elements of each are compatible.
+/// The irreducible polynomial `x^2 - 11` was chosen because `11` is
+/// the simplest choice of `B` for `x^2 - B` that makes this polynomial
+/// irreducible.
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:299]: The adjoined root squared (i.e. 11) is referred to as `BETA` elsewhere in the code, so let's do that in this comment as well.

```suggestion
/// Instances of `ExtElem` are elements of a finite field `F_p^2`. They are
/// represented as elements of `F_p[X] / (X^2 - 11)`. This large
/// finite field (about `2^128` elements) is used when the security of
/// operations depends on the size of the field. The field extension `ExtElem`
/// has `Elem` as a subfield, so operations on elements of each are compatible.
/// The irreducible polynomial `x^2 - 11` was chosen because `11` is
/// the simplest choice of `BETA` for `x^2 - BETA` that makes this polynomial
/// irreducible.
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_subtraction_easy() {
+        // 0 < a - b < P  (when a > b)
+        let fa: Elem = Elem::from(P - 2u64);
+        let fb: Elem = Elem::from(3u64);
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:626]:
```suggestion
        let fa = Elem::from(P - 2u64);
        let fb = Elem::from(3u64);
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+                fa + fb,
+                Elem::from(((a as u128 + b as u128) % P as u128) as u64),
+                "Error adding `{} + {}`",
+                a,
+                b
+            );
+            // This is a workaround that doesn't need to exist for baby bear
+            // because it doesn't overflow u64 under addition. Here, we could have P - b + a
+            // either overflow or wrap under u64.
+            let diff: u64 = if a < b {
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:724]:
```suggestion
            let diff = if a < b {
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 10]
```diff
@@ -0,0 +1,745 @@
...
+        for i in 0..self.0.len() {
+            self.0[i] += rhs.0[i];
+        }
+    }
+}
+
+impl ops::Sub for ExtElem {
+    type Output = Self;
+    /// Subtraction for [ExtElem]
+    fn sub(self, rhs: Self) -> Self {
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:429]:
```suggestion
    type Output = Self;
    
    /// Subtraction for [ExtElem]
    fn sub(self, rhs: Self) -> Self {
```

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+    const fn zero() -> Self {
+        Self::from_u64(0)
+    }
+
+    /// Returns the value one.
+    const fn one() -> Self {
+        Self::from_u64(1)
+    }
+
+    /// Returns the constant portion of a [Elem].
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:396]: IMO "constant portion" is unclear. The algorithm here is that given an element of the form `a + b * sqrt(BETA)`, return `a`. Maybe say "base field portion"?... open to other suggestions...

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+    pub fn const_part(self) -> Elem {
+        self.0[0]
+    }
+
+    /// Returns the elements of an [Elem].
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:401]: Similar wording confusion IMO. Maybe "Returns [Elem] as a vector of base field values"?... again open to suggestions.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+        // Test for addition when P < a + b < 2^64
+        let fa = Elem::from(P - 2u64);
+        let fb = Elem::from(5u64);
+        let a: u64 = fa.into();
+        let b: u64 = fb.into();
+        assert_eq!(fa + fb, Elem::from(a.wrapping_add(b)));
+    }
+
+    #[test]
+    fn compare_native_add_u64_wrap() {
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:613]: Am I missing something or is this the same test as `compare_native_add_subfield_wrap`?

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+            "Error multiplying `{} * {}`",
+            a,
+            b
+        );
+    }
+
+    #[test]
+    fn compare_native_multiplication_subfield_wrap() {
+        // P < a * b < 2^64
+        // We expect the result mod P, but there's no u64 overflow
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:691]: I think this does overflow u64, as (3/2) P > 2^64 - 1. Which is a reasonable test to have, it's just not what the comment says.

**Review by tzerrell — COMMENTED** [review]
Great! Looks good overall, I have some inline comments.

Several of my comments are from a documentation perspective, but I think the documentation needs a more thorough overhaul than just what's in my comments -- in particular aligning with & producing unified docs for the baby bear & the field element trait interface generally (i.e. the docs concerns you flagged in your initial comment on this PR). There is duplicated documentation and documentation that's worded as if the multiple fields and the trait interface don't exist as separate entities.

That said, I think that documentation cleanup should be a separate PR. We need it, we should do it promptly while this is fresh in your (and everyone's) mind, but IMO this PR doesn't need to depend on it.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 15]
```diff
@@ -0,0 +1,745 @@
...
+
+impl Default for Elem {
+    /// As a default, return the zero [Elem].
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+/// The modulus of our Goldilocks field: 2^64 - 2^32 + 1
+const P: u64 = (0u64.wrapping_sub(1) << 32) + 1;
```
Comment (jbruestle) [inline risc0/zkp/rust/src/field/goldilocks.rs:49]: The way this is computed maybe warrants a mild explaination?  Basically, the obvious thing of computing (1 << 64) - (1 << 32) + 1 would overflow U64, so we need to be trixy.

**Review by jbruestle — COMMENTED** [review]
Nice work, all the math seems right.  The 'compare_native' tests maybe should be renamed as they are more like 'check various edge cases by hand tests', or maybe like test vectors, but not sure what's a better name.

File: risc0/zkp/rust/src/field/goldilocks.rs (hunk 3 above)
Comment (3lkn) [reply risc0/zkp/rust/src/field/goldilocks.rs:87]: This one I'm not sure on, it's also in `baby_bear.rs` and I did not want to meddle in the affairs of macro wizards. @shkoo ?

Comment (3lkn) [reply risc0/zkp/rust/src/field/goldilocks.rs:87]: I am also not sure, it is in lines 97-100 of `baby_bear.rs` and I did not want to meddle in the affairs of macro wizards, @shkoo do we want this left in?

File: risc0/zkp/rust/src/field/goldilocks.rs (hunk 4 above)
Comment (3lkn) [reply risc0/zkp/rust/src/field/goldilocks.rs]: Minor note, the `let` in `let ret` on 278 causes new variable to be instantiated, switched it to `ret =` and kept the rest of this suggestion

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+macro_rules! rou_array {
+    [$($x:literal),* $(,)?] => {
+        [$(Elem::new($x)),* ]
+    }
+}
+
+/// Maximum power of two for which we have a root of unity using Goldilocks
+/// field
+impl field::RootsOfUnity for Elem {
+    const MAX_ROU_PO2: usize = 32;
```
Comment (tzerrell) [inline risc0/zkp/rust/src/field/goldilocks.rs:97]: To follow up on my earlier comment -- IMO pull this comment to just before the definition of `MAX_ROU_PO2`.
```suggestion
impl field::RootsOfUnity for Elem {
    /// Maximum power of two for which we have a root of unity using Goldilocks
    /// field
    const MAX_ROU_PO2: usize = 32;
```

**Review by tzerrell — COMMENTED** [review]
Looks good! IMO this just needs a resolution of the macro question and a few doc wording details.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 17]
```diff
@@ -0,0 +1,755 @@
...
+    }
+}
+
+/// The modulus of our Goldilocks field: 2^64 - 2^32 + 1
+/// Calculation steps chosen to avoid overflowing u64 with 2^64:
+/// 1. Wrapping subtract from 0u64, which leaves all ones
+/// 2. Left-shift ones over by 32, leaving 32 ones and 32 zeros (2^64-2^32)
+/// 3. Add one to get 2^64 - 2^32 + 1
+
+const P: u64 = (0u64.wrapping_sub(1) << 32) + 1;
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:54]:
```suggestion
const P: u64 = (0xffffffff_ffffffff << 32) + 1;
```

File: risc0/zkp/rust/src/field/baby_bear.rs [hunk 18]
```diff
@@ -12,15 +12,15 @@
...
-/// ! Support for the base finite field modulo 15*2^27 + 1
-use crate::field::{self, Elem as FieldElem};
-
 use alloc::fmt;
 use core::ops;
 
 use bytemuck::{Pod, Zeroable};
 
+/// ! Baby bear field.
+/// ! Support for the base finite field modulo 15*2^27 + 1
```
Comment (flaub) [inline risc0/zkp/rust/src/field/baby_bear.rs:21]: These module comments should probably appear before all the uses.

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 19]
```diff
@@ -0,0 +1,755 @@
...
+    // Compute shifted effect of medium
+    let med_shift = ((med as u64) << 32).wrapping_sub(med as u64);
+    // Add in, if overflow, subtract a P
+    let ret = ret.wrapping_add(med_shift);
+    let ret = if ret < med_shift || ret >= P {
+        ret.wrapping_sub(P)
+    } else {
+        ret
+    };
+    ret
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:292]:
```suggestion
    if ret < med_shift || ret >= P {
        ret.wrapping_sub(P)
    } else {
        ret
    }
```

File: risc0/zkp/rust/src/field/goldilocks.rs (hunk 3 above)
Comment (shkoo) [reply risc0/zkp/rust/src/field/goldilocks.rs:92]: This was the best way I could figure out to create a [Elem] from an array of uints in a const context.  I'd love a better way!

File: risc0/zkp/rust/src/field/goldilocks.rs [hunk 20]
```diff
@@ -0,0 +1,756 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::ops;
+
+use bytemuck::{Pod, Zeroable};
+
+/// ! Goldilocks field.
+/// ! Support for the base finite field modulo 2^64 - 2^32 + 1
```
Comment (flaub) [inline risc0/zkp/rust/src/field/goldilocks.rs:20]: Module comments (they should be in the form of `//! Some text`), should go just below the copyright notice.

File: risc0/zkp/rust/src/field/baby_bear.rs (hunk 18 above)
Comment (flaub) [reply risc0/zkp/rust/src/field/baby_bear.rs:21]: Also the format is: `//!`

**Comment by flaub** [comment]
LGTM

**Comment by 3lkn** [comment]
> Looks good! IMO this just needs a resolution of the macro question and a few doc wording details.

@tzerrell To my eyes the macro question is for-now-resolved, with this latest round of wording overhaul does this look ready to merge?

**Review by tzerrell — APPROVED** [review]

**Comment by tzerrell** [comment]
Let's do a followup PR for the outstanding comment cleanups to get the PRs landed in best merge conflict order
